internal_types/src/schema/merge.rs

```diff
@@ -0,0 +1,447 @@
+use super::{InfluxColumnType, Schema};
+use crate::schema::builder::set_field_metadata;
+use crate::schema::MEASUREMENT_METADATA_KEY;
+use arrow::datatypes::{DataType as ArrowDataType, Field, Schema as ArrowSchema};
+use hashbrown::hash_map::RawEntryMut;
+use hashbrown::HashMap;
+use snafu::Snafu;
+use std::sync::Arc;
+
+/// Database schema creation / validation errors.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("No schemas found when building merged schema"))]
+    NoSchemas,
+
+    #[snafu(display(
+        "Schema Merge Error: Incompatible measurement names. Existing measurement name '{}', new measurement name '{}'",
+        existing_measurement, new_measurement
+    ))]
+    TryMergeDifferentMeasurementNames {
+        existing_measurement: String,
+        new_measurement: String,
+    },
+
+    #[snafu(display(
+        "Schema Merge Error: Incompatible column type for '{}'. Existing type {:?}, new type {:?}",
+        field_name,
+        existing_column_type,
+        new_column_type
+    ))]
+    TryMergeBadColumnType {
+        field_name: String,
+        existing_column_type: Option<InfluxColumnType>,
+        new_column_type: Option<InfluxColumnType>,
+    },
+
+    #[snafu(display(
+        "Schema Merge Error: Incompatible data type for '{}'. Existing type {:?}, new type {:?}",
+        field_name,
+        existing_data_type,
+        new_data_type
+    ))]
+    TryMergeBadArrowType {
+        field_name: String,
+        existing_data_type: ArrowDataType,
+        new_data_type: ArrowDataType,
+    },
+
+    #[snafu(display(
+        "Schema Merge Error: Incompatible nullability for '{}'. Existing field {}, new field {}",
+        field_name, nullable_to_str(*existing_nullability), nullable_to_str(*new_nullability)
+    ))]
+    TryMergeBadNullability {
+        field_name: String,
+        existing_nullability: bool,
+        new_nullability: bool,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+fn nullable_to_str(nullability: bool) -> &'static str {
+    if nullability {
+        "can be null"
+    } else {
+        "can not be null"
+    }
+}
+
+/// Schema Merger
+///
+/// The usecase for merging schemas is when different chunks have
+/// different schemas. This struct can be used to build a combined
+/// schema by merging Schemas together according to the following
+/// rules:
+///
+/// 1. New columns may be added in subsequent schema, but the types of
+///    the columns (including any metadata) must be the same
+///
+/// 2. The measurement names must be consistent: one or both can be
+///    `None`, or they can both be `Some(name`)
+#[derive(Debug, Default)]
+pub struct SchemaMerger {
+    fields: HashMap<String, (Field, Option<InfluxColumnType>)>,
```

Comment:
It would help me read this code if there was a doc comment explaining how the keys and values should be interpreted. Perhaps something like:

```suggestion
    /// key: field name, values: field definition
    fields: HashMap<String, (Field, Option<InfluxColumnType>)>,
```